File: include/nol.php
<?php
namespace NOL;
class NolError extends \Exception {
    public $nolLine, $nolCol;
    public function __construct($msg, $line = 0, $col = 0) {
        parent::__construct("$msg at $line:$col");
        $this->nolLine = $line; $this->nolCol = $col;
    }
}
class NolParser {
    private $text, $pos = 0, $line = 1, $col = 1, $startTime, $depth = 0;
    public function __construct($text) { $this->text = $text; $this->startTime = microtime(true); }
    private function peek($n = 0) { return ($this->pos + $n < strlen($this->text)) ? $this->text[$this->pos + $n] : null; }
    private function advance() { $c = $this->peek(); $this->pos++; if ($c === "\n") { $this->line++; $this->col = 1; } else $this->col++; return $c; }
    private function skipWs() {
        while ($this->pos < strlen($this->text)) {
            $c = $this->text[$this->pos];
            if (ctype_space($c)) $this->advance();
            else if ($c === '#') {
                $this->advance();
                if ($this->peek() === '#') { $this->advance(); while ($this->pos < strlen($this->text) - 1 && !($this->text[$this->pos] === '#' && $this->text[$this->pos+1] === '#')) $this->advance(); if ($this->pos < strlen($this->text)) { $this->advance(); $this->advance(); } }
                else { while ($this->pos < strlen($this->text) && $this->text[$this->pos] !== "\n") $this->advance(); }
            } else break;
        }
    }
    private function readStr($q) {
        $s = "";
        while ($this->pos < strlen($this->text)) {
            $c = $this->peek(); if ($c === $q) { $this->advance(); break; }
            if ($c === '\\') { $this->advance(); $e = $this->advance(); if ($e === 'u') { $h = substr($this->text, $this->pos, 4); $this->pos += 4; $s .= mb_chr(hexdec($h)); } else if ($e === 'U') { $h = substr($this->text, $this->pos, 8); $this->pos += 8; $s .= mb_chr(hexdec($h)); } else $s .= ['n'=>"\n", 'r'=>"\r", 't'=>"\t"][$e] ?? $e; }
            else $s .= $this->advance();
        }
        return $s;
    }
    private function readKey() { $this->skipWs(); $c = $this->peek(); if ($c === '"' || $c === "'") return $this->readStr($this->advance()); $k = ""; while ($this->pos < strlen($this->text) && preg_match('/[a-zA-Z0-9_-]/', $this->text[$this->pos])) $k .= $this->advance(); return $k; }
    public function parseValue() {
        $this->skipWs(); if (microtime(true) - $this->startTime > 1) throw new NolError("Timeout", $this->line, $this->col); if (++$this->depth > 100) throw new NolError("Max depth", $this->line, $this->col);
        $c = $this->peek(); $res = null;
        if ($c === '{') { $this->advance(); $res = []; while ($this->pos < strlen($this->text)) { $this->skipWs(); if ($this->peek() === '}') { $this->advance(); break; } $this->parsePair($res); $this->skipWs(); if ($this->peek() === ',') $this->advance(); } }
        else if ($c === '[') { $this->advance(); $res = []; while ($this->pos < strlen($this->text)) { $this->skipWs(); if ($this->peek() === ']') { $this->advance(); break; } $res[] = $this->parseValue(); $this->skipWs(); if ($this->peek() === ',') $this->advance(); } }
        else if ($c === '"' || $c === "'") $res = $this->readStr($this->advance());
        else if (preg_match('/[0-9-]/', $c)) { $b = ""; while ($this->peek() && preg_match('/[0-9.eE+-]/', $this->peek())) $b .= $this->advance(); $res = (strpos($b, '.') !== false || stripos($b, 'e') !== false) ? (float)$b : (int)$b; }
        else { $b = ""; while ($this->peek() && ctype_alpha($this->peek())) $b .= $this->advance(); if ($b === "true") $res = true; else if ($b === "false") $res = false; else if ($b === "null") $res = null; else throw new NolError("Invalid: $b", $this->line, $this->col); }
        $this->depth--; return $res;
    }
    public function parsePair(&$obj) { $this->skipWs(); $key = $this->readKey(); if (class_exists('Normalizer')) $key = \Normalizer::normalize($key, \Normalizer::FORM_C); if (in_array($key, ["_env", "_interpolate", "_meta"])) throw new NolError("Reserved: $key", $this->line, $this->col); $this->skipWs(); if ($this->advance() !== ':') throw new NolError("Expected :", $this->line, $this->col); $val = $this->parseValue(); if (isset($obj[$key])) throw new NolError("Duplicate: $key", $this->line, $this->col); $obj[$key] = $val; }
    private function parseInto(&$obj) { while ($this->pos < strlen($this->text)) { $this->skipWs(); if (!$this->peek() || $this->peek() === '[') break; $this->parsePair($obj); } }
    public function parse() { $root = []; while ($this->pos < strlen($this->text)) { $this->skipWs(); if (!$this->peek()) break; if ($this->peek() === '[') { $this->advance(); $path = ""; while ($this->peek() && $this->peek() !== ']') $path .= $this->advance(); if ($this->peek() === ']') $this->advance(); $parts = explode('.', $path); $curr = &$root; foreach ($parts as $i => $p) { if (!isset($curr[$p])) $curr[$p] = []; $curr = &$curr[$p]; } $this->parseInto($curr); } else $this->parsePair($root); } return $root; }
}
function parse($text) { return (new NolParser($text))->parse(); }
function dump($v, $indent = 2, $level = 0, $root = false) {
    if ($v === null) return "null"; if (is_bool($v)) return $v ? "true" : "false";
    if (is_numeric($v)) { $s = sprintf("%.5f", $v); while (str_contains($s, '.') && str_ends_with($s, '0')) $s = substr($s, 0, -1); if (str_ends_with($s, '.')) $s .= '0'; return $s; }
    if (is_string($v)) return $root ? $v : "\"$v\"";
    if (array_is_list($v)) return "[" . implode(", ", array_map(fn($x) => dump($x, $indent, $level + 1), $v)) . "]";
    ksort($v); $pad = str_repeat(" ", $level * $indent); $nextPad = str_repeat(" ", ($level + 1) * $indent);
    $items = []; foreach ($v as $k => $val) { if (str_starts_with($k, "_") && !$root) continue; $items[] = $root ? "$k: " . dump($val, $indent, $level + 1) : "\n$nextPad$k: " . dump($val, $indent, $level + 1); }
    if ($root) return implode("\n", $items); if (empty($items)) return "{}"; return "{" . implode(",", $items) . "\n$pad}";
}

File: include/nol.rs
use std::collections::{BTreeMap};
use std::time::{Instant, Duration};

#[derive(Debug, Clone, PartialEq)]
pub enum NolValue {
    Null, Bool(bool), Int(i64), Float(f64), String(String), Array(Vec<NolValue>), Object(BTreeMap<String, NolValue>),
}

impl NolValue {
    pub fn dump(&self, indent: usize, level: usize, is_root: bool) -> String {
        match self {
            NolValue::Null => "null".to_string(),
            NolValue::Bool(b) => b.to_string(),
            NolValue::Int(i) => i.to_string(),
            NolValue::Float(f) => { let mut s = format!("{:.5}", f); while s.ends_with('0') { s.pop(); } if s.ends_with('.') { s.push('0'); } s }
            NolValue::String(s) => if is_root { s.clone() } else { format!("{:?}", s) },
            NolValue::Array(a) => format!("[{}]", a.iter().map(|v| v.dump(indent, level + 1, false)).collect::<Vec<_>>().join(", ")),
            NolValue::Object(o) => {
                let pad = " ".repeat(level * indent); let next_pad = " ".repeat((level + 1) * indent);
                let mut lines = Vec::new();
                for (k, v) in o { if k.starts_with('_') { continue; } let val = v.dump(indent, level + 1, false); if is_root { lines.push(format!("{}: {}", k, val)); } else { lines.push(format!("\n{}{}: {}", next_pad, k, val)); } }
                if is_root { return lines.join("\n"); } if o.is_empty() { return "{}".to_string(); }
                format!("{{{}\n{}}}", lines.join(","), pad)
            }
        }
    }
}

pub struct Document {
    pub root: BTreeMap<String, NolValue>,
}

impl Document {
    pub fn get(&self, path: &str) -> Option<&NolValue> {
        let parts: Vec<&str> = path.split('.').collect();
        let mut curr = &self.root;
        for (i, p) in parts.iter().enumerate() {
            if i == parts.len() - 1 {
                return curr.get(*p);
            } else {
                match curr.get(*p) {
                    Some(NolValue::Object(o)) => curr = o,
                    _ => return None,
                }
            }
        }
        None
    }
    pub fn exists(&self, path: &str) -> bool { self.get(path).is_some() }
}

pub struct Builder {
    pub root: BTreeMap<String, NolValue>,
}

impl Builder {
    pub fn new() -> Self { Self { root: BTreeMap::new() } }
    pub fn set(mut self, path: &str, value: NolValue) -> Self {
        let parts: Vec<String> = path.split('.').map(|s| s.to_string()).collect();
        let mut curr = &mut self.root;
        for p in parts.iter().take(parts.len() - 1) {
            if !curr.contains_key(p) || !matches!(curr.get(p), Some(NolValue::Object(_))) {
                curr.insert(p.clone(), NolValue::Object(BTreeMap::new()));
            }
            curr = match curr.get_mut(p).unwrap() {
                NolValue::Object(o) => o,
                _ => unreachable!(),
            };
        }
        curr.insert(parts.last().unwrap().clone(), value);
        self
    }
    pub fn build(self) -> Document { Document { root: self.root } }
}

pub struct NolParser<'a> { text: &'a str, pos: usize, start: Instant, depth: usize }
impl<'a> NolParser<'a> {
    pub fn new(text: &'a str) -> Self { Self { text, pos: 0, start: Instant::now(), depth: 0 } }
    fn peek(&self) -> Option<char> { self.text[self.pos..].chars().next() }
    fn advance(&mut self) -> Option<char> { let c = self.peek()?; self.pos += c.len_utf8(); Some(c) }
    fn skip_ws(&mut self) { while let Some(c) = self.peek() { if c.is_whitespace() { self.advance(); } else if c == '#' { self.advance(); if self.peek() == Some('#') { self.advance(); while self.pos < self.text.len() - 1 && &self.text[self.pos..self.pos+2] != "##" { self.advance(); } if self.pos < self.text.len() { self.advance(); self.advance(); } } else { while let Some(c) = self.peek() { if c == '\n' { break; } self.advance(); } } } else { break; } } }
    fn parse_hex(&mut self, n: usize) -> char { let mut h = String::new(); for _ in 0..n { h.push(self.advance().expect("Hex")); } std::char::from_u32(u32::from_str_radix(&h, 16).expect("HexVal")).expect("Char") }
    fn read_str(&mut self, q: char) -> String { let mut s = String::new(); while let Some(c) = self.peek() { if c == q { self.advance(); return s; } if c == '\\' { self.advance(); let e = self.advance().expect("Esc"); if e == 'u' { s.push(self.parse_hex(4)); } else if e == 'U' { s.push(self.parse_hex(8)); } else { s.push(match e { 'n'=>'\n', 'r'=>'\r', 't'=>'\t', x=>x }); } } else { s.push(self.advance().unwrap()); } } s }
    fn read_key(&mut self) -> String { self.skip_ws(); let c = self.peek().expect("Key"); if c == '"' || c == '\'' { let q = self.advance().unwrap(); self.read_str(q) } else { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_alphanumeric() || c == '_' || c == '-' { s.push(self.advance().unwrap()); } else { break; } } s } }
    pub fn parse_value(&mut self) -> NolValue {
        self.skip_ws(); if self.start.elapsed() > Duration::from_secs(1) { panic!("Timeout"); } self.depth += 1; if self.depth > 100 { panic!("Depth"); }
        let c = self.peek().expect("EOF");
        let res = match c {
            '{' => { self.advance(); let mut o = BTreeMap::new(); while let Some(_) = self.peek() { self.skip_ws(); if self.peek() == Some('}') { self.advance(); break; } self.parse_pair(&mut o); self.skip_ws() ; if self.peek() == Some(',') { self.advance(); } } NolValue::Object(o) }
            '[' => { self.advance(); let mut a = Vec::new(); while let Some(_) = self.peek() { self.skip_ws(); if self.peek() == Some(']') { self.advance(); break; } a.push(self.parse_value()); self.skip_ws(); if self.peek() == Some(',') { self.advance(); } } NolValue::Array(a) }
            '"' | '\'' => { let q = self.advance().unwrap(); NolValue::String(self.read_str(q)) }
            '0'..='9' | '-' => { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_ascii_digit() || ".-eE+".contains(c) { s.push(self.advance().unwrap()); } else { break; } } if s.contains('.') || s.to_lowercase().contains('e') { NolValue::Float(s.parse().unwrap()) } else { NolValue::Int(s.parse().unwrap()) } }
            _ => { let mut s = String::new(); while let Some(c) = self.peek() { if c.is_alphabetic() { s.push(self.advance().unwrap()); } else { break; } } match s.as_str() { "true" => NolValue::Bool(true), "false" => NolValue::Bool(false), "null" => NolValue::Null, _ => panic!("Invalid: {}", s) } }
        }; self.depth -= 1; res
    }
    fn parse_pair(&mut self, o: &mut BTreeMap<String, NolValue>) {
        self.skip_ws(); let key = self.read_key(); if ["_env", "_interpolate", "_meta"].contains(&key.as_str()) { panic!("Reserved: {}", key); }
        self.skip_ws(); if self.advance() != Some(':') { panic!("Expected : for {}", key); }
        let val = self.parse_value(); if o.contains_key(&key) { panic!("Duplicate: {}", key); } o.insert(key, val);
    }
    pub fn parse(&mut self) -> Document {
        let mut root = BTreeMap::new();
        while self.pos < self.text.len() {
            self.skip_ws(); if self.peek().is_none() { break; }
            if self.peek() == Some('[') {
                self.advance(); let mut path_s = String::new(); while let Some(c) = self.peek() { if c == ']' { self.advance(); break; } path_s.push(self.advance().unwrap()); }
                let parts: Vec<String> = path_s.split('.').map(|s| s.to_string()).collect();
                let mut curr = &mut root;
                for p in parts.iter().take(parts.len() - 1) {
                    if !curr.contains_key(p) || !matches!(curr.get(p), Some(NolValue::Object(_))) {
                        curr.insert(p.clone(), NolValue::Object(BTreeMap::new()));
                    }
                    curr = match curr.get_mut(p).unwrap() {
                        NolValue::Object(o) => o,
                        _ => unreachable!(),
                    };
                }
                let last_p = parts.last().unwrap();
                if !curr.contains_key(last_p) || !matches!(curr.get(last_p), Some(NolValue::Object(_))) {
                    curr.insert(last_p.clone(), NolValue::Object(BTreeMap::new()));
                }
                if let NolValue::Object(ref mut o) = curr.get_mut(last_p).unwrap() {
                    self.parse_into(o);
                }
            } else { self.parse_pair(&mut root); }
        } Document { root }
    }
    fn parse_into(&mut self, o: &mut BTreeMap<String, NolValue>) { while self.pos < self.text.len() { self.skip_ws(); match self.peek() { Some('[') | None => break, _ => self.parse_pair(o) } } }
}
pub fn parse(text: &str) -> Document { let mut p = NolParser::new(text); p.parse() }
